File: workspace/test/adapter/http_greeting_api_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:saying_hello/adapter/http_greeting_api.dart';
import 'package:test/test.dart';

void main() {
  test('requests the canonical greeting endpoint', () async {
    final client = MockClient((request) async {
      expect(
        request.url.toString(),
        'http://localhost:25664/api/greeting?name=Ada',
      );

      return http.Response(jsonEncode({'message': 'Hello, Ada!'}), 200);
    });

    final api = HttpGreetingApi(
      baseUrl: 'http://localhost:25664',
      client: client,
    );
    final result = await api.getGreeting('Ada');

    expect(result.message, 'Hello, Ada!');
  });
}

File: workspace/src/adapter/http-greeting-api.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import { HttpGreetingApi } from "./http-greeting-api";

describe("HttpGreetingApi", () => {
  afterEach(() => {
    vi.unstubAllGlobals();
  });

  it("requests the canonical greeting endpoint", async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      json: vi.fn().mockResolvedValue({ message: "Hello, Ada!" }),
    });
    vi.stubGlobal("fetch", fetchMock);

    const api = new HttpGreetingApi("http://localhost:25616");
    const result = await api.getGreeting("Ada");

    expect(fetchMock).toHaveBeenCalledWith(
      "http://localhost:25616/api/greeting?name=Ada",
    );
    expect(result).toEqual({ message: "Hello, Ada!" });
  });
});
